File: Graphs/BFS.c
// input adjacency matrix

#include <stdio.h>
int k=0; // indexing for the final bfs array
int n;
int f=0,r=-1; //queue front and rear
void BFS(int g[n][n],int i,int *visited,int *order,int *queue)
{
    if(f>r) return;
    visited[i]=1;
    order[k++] = i+1;
    for(int j=0;j<n;j++)
    {
        if(g[i][j]==1 && visited[j]==0)
        {
            visited[j]=1;
            queue[++r]=j+1;
        }
    }
    f++;
    BFS(g,queue[f]-1,visited,order,queue);
}

void main() 
{
	printf("No of vertices\n");
	scanf("%d",&n);
	int graph[n][n],queue[n];
	printf("Enter Adj. Matrix\n");
	for(int i=0;i<n;i++)
	    for(int j=0;j<n;j++)
	        scanf("%d",&graph[i][j]);
	int visited[n];
    for(int i=0;i<n;i++)
    {
        queue[i]=0;
        visited[i]=0;
    }
    int order[n];
    int x;
    printf("Starting vertex(Naming starts with 1,2,3...)\n");
    scanf("%d",&x);
    queue[++r]=x;
	BFS(graph,x-1,visited,order,queue);
	for(int i=0;i<n;i++)
	    printf("%d ",order[i]);
}

File: Graphs/BFS.java
import java.io.*;
import java.util.*;

class Graph{
    int V; //number of vertices
    LinkedList<Integer> adj[];
    
    Graph(int v){  //not a default constructor
        V = v; //initializing
        adj = new LinkedList[v];
        // create...
        for(int i=0;i<v;i++){
            adj[i] = new LinkedList();
        }
    }
    
    //add edges
    void addEdge(int v,int w){ // w to be added
        adj[v].add(w);
    }
    
    // s is starting vertex(source)
    // BFS traversal
    
    void BFS(int s){
        boolean []visited = new boolean[V]; // create visited list for all vertices
        // if vertex is visited mark as 'true'
        
        LinkedList<Integer> queue = new LinkedList<Integer>(); //use LL class as queue(since bfs)
        //mark starting vertex as visited and add to queue
        visited[s] = true;
        queue.add(s);
        
        //BFS algorithm
        while(queue.size()!=0){
            s = queue.poll(); // removes first element, store in s
            //print it
            System.out.print(s+" ");
            //use Iterator class to get adjacent vertices
            Iterator<Integer> i = adj[s].listIterator();
            while(i.hasNext()){
                int n = i.next();
                if(visited[n] == false){
                    visited[n] = true;
                    queue.add(n);
                }
            }
        }
    }
    //main 
    public static void main (String[] args) {
        Scanner s = new Scanner(System.in);
        Graph g = new Graph(4);
        g.addEdge(0,1);
        g.addEdge(0,2); 
        g.addEdge(1,2); 
        g.addEdge(2,0); 
        g.addEdge(2,3); 
        g.addEdge(3,3);
        int k;
        System.out.println("Enter the starting vertex/ source");
        k = s.nextInt();
        
        System.out.print("BFS Traversal is ");
        
        g.BFS(k);
    }
}
